nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/blob/TransactionalBlobStore.java

```diff
@@ -0,0 +1,362 @@
+import org.nuxeo.ecm.core.blob.binary.BinaryGarbageCollector;
+import org.nuxeo.runtime.jtajca.NuxeoContainer;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+
+/**
+ * Transactional Blob Store.
+ * <p>
+ * Until the transaction is committed, blobs are stored in a transient store. Upon commit, they are sent to the
+ * permanent store.
+ * <p>
+ * It is important that a copy operation between the transient store and the permanent store be extremely fast and never fail, as it
+ * will be done during commit.
+ *
+ * @since 11.1
+ */
+public class TransactionalBlobStore extends AbstractBlobStore implements Synchronization {
+
+    private static final Log log = LogFactory.getLog(TransactionalBlobStore.class);
+
+    public final BlobStore store;
+
+    // may be the same as the permanent store if it has versioning
+    public final BlobStore transientStore;
+
+    /**
+     * Transient data recording operations applied to a blob, to be executed on the permanent store at commit time.
+     */
+    public static class TransientInfo {
+
+        /** The key in the transient store of the blob to use, or a delete marker. */
+        public String transientKey;
+
+        /** The update to apply. */
+        public BlobUpdateContext blobUpdateContext;
+    }
+
+    protected final ThreadLocal<Map<String, TransientInfo>> transientInfo = new ThreadLocal<>();
+
+    // the keys that have been modified in any active transaction
+    protected final Map<String, Transaction> keysInActiveTransactions = new ConcurrentHashMap<>();
+
+    protected static final String DELETE_MARKER = "";
+
+    protected static boolean isDeleteMarker(String transientKey) {
+        return DELETE_MARKER.equals(transientKey);
+    }
+
+    public TransactionalBlobStore(BlobStore store, BlobStore transientStore) {
+        super("tx", store.getKeyStrategy());
+        this.store = store;
+        this.transientStore = transientStore;
+        if (store.hasVersioning() && transientStore != store) {
+            throw new NuxeoException("If the store has versioning then it must be also the transient store");
+        }
+    }
+
+    @Override
+    public BinaryGarbageCollector getBinaryGarbageCollector() {
+        return store.getBinaryGarbageCollector();
+    }
+
+    @Override
+    public boolean hasVersioning() {
+        return store.hasVersioning();
+    }
+
+    @Override
+    public BlobStore unwrap() {
+        return store.unwrap();
+    }
+
+    @Override
+    public String writeBlob(BlobWriteContext blobWriteContext)
+            throws IOException {
+        if (TransactionHelper.isTransactionActive()) {
+            String transientKey;
+            String key;
+            logTrace("group tx write");
+            if (hasVersioning()) {
+                // with versioning there can be no collisions, and transientStore = store
+                transientKey = transientStore.writeBlob(blobWriteContext);
+                key = transientKey;
```

Comment:
perhaps i am missed something. But the idea behind this transactional blob provider is to be transactional with the `TX` commit on Nuxeo side. is it right ? because in a case where we have versioning on blob store activated like S3, the `transient store` will be the same that the final one `S3` and we will write directly the blob without waiting the transaction commit.

Reply:
Yes, when S3 versioning is active we write the blob immediately and report its final key (that includes the S3 version id) because it cannot collide with another blob write (which would have a different version id and therefore a different key).